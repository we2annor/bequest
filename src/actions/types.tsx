export const ADD_CONTACT = 'ADD_CONTACT';
export const FETCH_CONTACTS = 'FETCH_CONCTACTS';
export const FETCH_CONTACT = 'FETCH_CONTACT';
export const DELETE_CONTACT = 'DELETE_CONTACT';
export const EDIT_CONTACT = 'EDIT_CONTACT';

export interface AddressesState {
    addressBook: Address;
    //id?:ID
}

export interface Address {
    houseNo: string;
    streetName: string;
    locality: string;
    postTown: string;
    county: string;
    postCode: string;
    id: number
}

interface ID {
    id: number
}

export type AddressKey = keyof Address;

export const InitialAddressValues = {
    houseNo:'',
    streetName:'',
    locality: '',
    postTown : '',
    county: '',
    postCode: '',
    id: 0
}

export interface AddressFormProps {
  selectedAddress: Address;
  postcode: string;
  addContact: (formValues:Address)=>void;
}

export interface AddNewContact {
    type: typeof ADD_CONTACT
    payload: Address
}

export interface FetchContact {
    type: typeof FETCH_CONTACT;
    payload: Address
}

export interface FetchContacts {
    type: typeof FETCH_CONTACTS;
    payload: Address
}

export interface EditContact {
    type: typeof EDIT_CONTACT;
    payload: Address
}

export interface DeleteContact {
    type : typeof DELETE_CONTACT;
    payload: ID
}

export type AddressBookActionTypes = AddNewContact | FetchContact | FetchContacts | EditContact | DeleteContact ; 