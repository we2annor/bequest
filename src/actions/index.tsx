import { Dispatch } from "redux";
import {
  ADD_CONTACT,
  FETCH_CONTACT,
  FETCH_CONTACTS,
  EDIT_CONTACT,
  DELETE_CONTACT,
  Address,
} from "./types";
import { addressBook } from "../api/addressBook";
import history from "../components/history";

export const addContact = (formValues: Address) => async (
  dispatch: Dispatch
) => {
  const response = await addressBook.post("/addressBook", formValues);
  dispatch({ type: ADD_CONTACT, payload: response.data });
  console.log("add contact", formValues);
  history.push("/");
};

export const fetchContact = (id: number) => async (dispatch: Dispatch) => {
  const response = await addressBook.get(`/addressBook/${id}`);
  dispatch({
    type: FETCH_CONTACT,
    payload: response.data,
  });
};

export const editContact = (id: number) => async (dispatch: Dispatch) => {
  const response = await addressBook.get(`/addressBook/${id}`);
  dispatch({
    type: EDIT_CONTACT,
    payload: response.data,
  });
};

export const deleteContact = (id: any) => async (dispatch: Dispatch) => {
  await addressBook.get(`/addressBook/${id}`);
  dispatch({ type: DELETE_CONTACT, payload: id });
};

export const fetchContacts = () => async (dispatch: Dispatch) => {
  const response = await addressBook.get("/addressBook");
  dispatch({
    type: FETCH_CONTACTS,
    payload: response.data,
  });
};
