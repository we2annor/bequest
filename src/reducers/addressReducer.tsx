import _ from "lodash";
import {
  ADD_CONTACT,
  FETCH_CONTACT,
  FETCH_CONTACTS,
  EDIT_CONTACT,
  //DELETE_CONTACT,
  AddressesState,
  AddressBookActionTypes,
} from "../actions/types";

const initialState : AddressesState = Object({})

export const addressReducer = (
  state = initialState,
  action: AddressBookActionTypes
): AddressesState => {
  switch (action.type) {
    case FETCH_CONTACT:
      return { ...state, [action.payload.id]: action.payload };
    case FETCH_CONTACTS:
      return { ...state, ..._.mapKeys(action.payload, "id") };
    case ADD_CONTACT:
      return { ...state, [action.payload.id]: action.payload };
    case EDIT_CONTACT : 
      return {...state, [action.payload.id]: action.payload};
    //case DELETE_CONTACT: 
      //return _.omit(state, action.payload);
    default:
      return state;
  }
};
