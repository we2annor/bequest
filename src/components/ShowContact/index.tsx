import React, {useEffect} from 'react';
import { connect } from 'react-redux';
import { fetchContact } from '../../actions';
import { Address } from '../../actions/types';

interface IProps {
    contact: Address;
    fetchContact: (id:number)=> void;
    match: {
        params: {
            id: number;
        }
    };
}

const ShowContact:React.FC<IProps> =({contact, fetchContact, match})=>{
    useEffect(()=>{
        const fetchSelectedAddress =()=>{
            const {id} = match.params;
            fetchContact(id);
        }
        fetchSelectedAddress();
    },[fetchContact, match.params]);

    if(!contact){
        return <div>No contact</div>;
    }
    
    return (
        <div className="contact">
            <h2>Contact Id: {contact.id}</h2>
            <div className="contact__list">
                <div className="list-item"><span>Full name :</span>{contact.firstName} {contact.lastName}</div>
                <div className="list-item"><span>Address :</span>
                    {contact.houseNo}, 
                    {contact.streetName}, 
                    {contact.locality}
                </div>
                <div className="list-item"><span>City :</span>{contact.postTown}</div>
                <div className="list-item"><span>Country:</span>{contact.country}</div>
                <div className="list-item"><span>Post code:</span>{contact.postCode}</div>
            </div>
        </div>
    )
}

const mapPropsToState =(state: any, ownProps: any)=>{
    return { contact : state.addressBook[ownProps.match.params.id]}
}

export default connect(mapPropsToState, {fetchContact})(ShowContact);

