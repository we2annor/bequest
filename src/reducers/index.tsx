import {combineReducers} from 'redux';
import { addressReducer } from './addressReducer';

export const rootReducer = combineReducers({
    addressBook: addressReducer
})

export type RootState = ReturnType< typeof rootReducer>