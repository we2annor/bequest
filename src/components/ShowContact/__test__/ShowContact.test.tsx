import React from "react";
import { render, cleanup } from "@testing-library/react";
import { Provider } from "react-redux";
import { createStore, applyMiddleware } from "redux";
import thunk from "redux-thunk";
import renderer from "react-test-renderer";
import { rootReducer } from "../../../reducers";

import ShowContact from "../index";
import { Address } from "../../../actions/types";

afterEach(cleanup);

const contact: Address = {
  firstName: "Ernest",
  lastName: "Annor",
  houseNo: "123",
  streetName: "new street",
  locality: "wembley",
  country: "London",
  postCode: "ha9 2gm",
  postTown: "Middlesex",
  id: 0,
};

const match = {
  params: {
    id: 1,
  },
};

const store = createStore(rootReducer, applyMiddleware(thunk));

describe("When ShowContact Mounts", () => {
  test("Should Render successfully", () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <ShowContact contact={contact} match={match} />
      </Provider>
    );

    expect(getByTestId("contact")).toBeTruthy();
  });

  test("Matches snapshot", () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <ShowContact contact={contact} match={match} />
        </Provider>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
