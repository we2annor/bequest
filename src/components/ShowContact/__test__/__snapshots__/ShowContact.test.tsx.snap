// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When ShowContact Mounts Matches snapshot 1`] = `
<div>
  No contact
</div>
`;
